### DISCUSSION TITLE
About rawValueWrapper state initialization

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#119] [General]
Having this:
```dart
/// A collection of experimental side effects to
/// complement the [BuiltinSideEffects].
extension ExperimentalSideEffects on SideEffectRegistrar {
  /// Returns a raw value wrapper; i.e., a getter and setter for some value.
  /// *The setter will not trigger rebuilds*.
  /// The initial state will be set to the result of running [init],
  /// if it was provided. Otherwise, you must manually set it
  /// via the setter before ever calling the getter.
  // NOTE: experimental because I am not sold on the API and use cases
  // for non-internal usage.
  (T Function(), void Function(T)) rawValueWrapper<T>([T Function()? init]) {
    return use.callonce(() {
      late T state;
      if (init != null) state = init();
      return (() => state, (T newState) => state = newState);
    });
  }
}
```

Wouldn't be desirable to delay the lazy initialization to the moment where the getter is executed first?

My use case is a "module manager" that allows a project to override any library module, so if lazy initialization would be as I suggest it would prevent to create the library module when it is override by the project:
```dart
extension _ConvenienceExtension on SideEffectRegistrar {
  SideEffectRegistrar get use => this;
}

extension ModulesSideEffectRegistrarExtension on SideEffectRegistrar {
  ModuleManager<T> asModuleManager<T extends Module>(
      T Function() defaultModuleBuilder) {
    final (getter, setter) = use.rawValueWrapper(defaultModuleBuilder);
    final (getState, setState) =
        use.rawValueWrapper(() => ModuleManagerState.defaultModule);

    return ModuleManager(
      get: getter,
      set: (module) {
        if (getState() == ModuleManagerState.overrideModule) {
          throw ('ModulesSideEffectRegistrarExtension.asModuleManager() -- $T -- Bad usage -- Already overrided');
        }

        setter(module);

        setState(ModuleManagerState.overrideModule);
      },
    );
  }
}

class ModuleManager<T extends Module> {
  final T Function() get;
  final void Function(T) set;

  ModuleManager({
    required this.get,
    required this.set,
  });
}

enum ModuleManagerState {
  defaultModule,
  overrideModule,
}

```

**Comment by busslina** [comment]
\* You explained me how to override Capsules, and I use that pattern. But for this particular case I want to test this other way that will simplify to me to override `n` modules.

**Comment by GregoryConrad** [comment]
What you're doing right now feels kinda clunky. `rawValueWrapper` should be a pretty rare side effect in normal code, and I wouldn't normally recommend using it since it is experimental (I reserve the right to remove/modify it in any release). I don't have a strong enough reason to change it right now or in the foreseeable future, but just know that it could happen.

Now, based on what I'm seeing, you might just want to use the [`use.lazyStateGetterSetter`](https://pub.dev/documentation/rearch/latest/rearch/BuiltinSideEffects/lazyStateGetterSetter.html) side effect. It'll give you a `T Function() get, void Function(T) set`, which seems to be what you're after. If you really want the "have I been set before" check, you might be able to do something like:
```dart
final (getter, setter) = use.lazyStateGetterSetter(() => defaultModule);
return (getter, (module) {
  if (getter() != defaultModule) {
    // throw your only can be set once error
  }
  setter(module);
});
```

**Reply by GregoryConrad** [reply]
that will always result in one call to the `() => defaultModule` (or whatever your init function is), though. If that is a problem then lmk

**Reply by busslina** [reply]
But it seems that `lazyStateGetterSetter` also initialize the state at first capsule read instead of at first `getter` execution. So, if I read the capsule in order to override the "module" by calling `set` (which I will refactor its parameter from `T` to `T Function()`), the default module will be also built/created, and this is what I want to prevent.


```dart
/// Side effect that provides a way for capsules to contain some state,
  /// where the initial state is computationally expensive.
  /// Further, instead of returning the state directly, this instead returns
  /// a getter that is safe to capture in closures.
  /// Similar to the `useState` hook from React;
  /// see https://react.dev/reference/react/useState
  (T Function(), void Function(T)) lazyStateGetterSetter<T>(T Function() init) {
    // We use register directly to keep the same setter function across rebuilds
    // (but we need to return a new getter on each build, see below for more)
    final (getter, setter) = use.register((api) {
      var state = init();

      T getter() => state;
      void setter(T newState) {
        api.rebuild((cancelRebuild) {
          if (newState == state) {
            cancelRebuild();
            return;
          }

          state = newState;
        });
      }

      return (getter, setter);
    });
```

**Reply by busslina** [reply]
I know that this initialization is not too expensive in this case, but I think it will be good to have a mechanism to delay it to the `getter` execution. Maybe I can have a workaround by changing `lazyStateGetterSetter<T>` to `lazyStateGetterSetter<T Function()>`.

But, what do you think on having a ReArch built-in option to handle this?

**Reply by GregoryConrad** [reply]
> the default module will be also built/created, and this is what I want to prevent.

ahh, ok. If lazyStateGetterSetter delayed initialization, would that otherwise solve your use case with the module stuff? Or do you need something else too?

**Reply by busslina** [reply]
Yes, it would be enough. But maybe you want to have `lazyStateGetterSetter` like it is now and to add an alternative method that does the job

**Reply by GregoryConrad** [reply]
I think I’d rather modify the lazyGetterSetter; lazy is in its name, so I think it’d intuitively make sense to not be initialized until it’s actually read. Do you have a situation where you’d always want to call the init function on the first build that currently uses the lazy getter/setter?

**Reply by busslina** [reply]
> Do you have a situation where you’d always want to call the init function on the first build that currently uses the lazy getter/setter?

Not really :)

> lazy is in its name, so I think it’d intuitively make sense to not be initialized until it’s actually read

Yeah, it make sense

**Reply by GregoryConrad** [reply]
https://github.com/GregoryConrad/rearch-dart/commit/127ff36810498bf1bbc04c5fa9bc8a4310f7cf37

Will be released shortly

**Comment by busslina** [comment]
I removed all my `rawValueWrapper` usages (combined with the use of `use.transactionRunner()`):
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/54be80df-bd70-404e-8fdb-853c3362afdd)
